modules/misc/listenbrainz.c

```diff
@@ -0,0 +1,522 @@
+    free (p_sys);
+    return 0;
+}
+
+static int Open (vlc_object_t *p_this)
+{
+    intf_thread_t *p_intf = ( intf_thread_t * ) p_this;
+    intf_sys_t *p_sys = calloc (1 , sizeof (intf_sys_t));
+
+    if (!p_sys)
+        return VLC_ENOMEM;
+
+    p_intf->p_sys = p_sys;
+
+    if (!Configure (p_intf))
+        return VLC_EGENERIC;
+
+    static struct vlc_player_cbs const player_cbs =
+        {
+            .on_state_changed = PlayerStateChanged ,
+            .on_current_media_changed = OnCurrentMediaChanged ,
+        };
+    static struct vlc_player_timer_cbs const timer_cbs =
+        {
+            .on_update = OnTimerUpdate ,
+            .on_discontinuity = OnTimerStopped ,
+        };
+
+    vlc_playlist_t *playlist = vlc_intf_GetMainPlaylist (p_intf);
+    p_sys->player = vlc_playlist_GetPlayer (playlist);
+
+    vlc_player_Lock (p_sys->player);
+    p_sys->player_listener = vlc_player_AddListener (p_sys->player , &player_cbs , p_intf);
+    vlc_player_Unlock (p_sys->player);
+
+    if (!p_sys->player_listener)
+        goto error;
+
+    p_sys->timer_listener = vlc_player_AddTimer (p_sys->player , VLC_TICK_FROM_SEC (1) , &timer_cbs , p_intf);
+    if (!p_sys->timer_listener)
+        goto error;
+
+    vlc_mutex_init (&p_sys->lock);
+    vlc_cond_init (&p_sys->wait);
+
+    if (vlc_clone (&p_sys->thread , Run , p_intf , VLC_THREAD_PRIORITY_LOW))
+        goto error;
+
+    return VLC_SUCCESS;
+
+error:
+    vlc_player_Lock (p_sys->player);
+    if (p_sys->player_listener)
+        vlc_player_RemoveListener (p_sys->player , p_sys->player_listener);
+    if (p_sys->timer_listener)
+        vlc_player_RemoveTimer (p_sys->player , p_sys->timer_listener);
+    vlc_player_Unlock (p_sys->player);
+    free (p_sys);
+    return VLC_EGENERIC;
+}
+
+static void Close (vlc_object_t *p_this)
+{
+    intf_thread_t *p_intf = ( intf_thread_t * ) p_this;
+    intf_sys_t *p_sys = p_intf->p_sys;
+
+    vlc_cancel (p_sys->thread);
+    vlc_join (p_sys->thread , NULL);
+
+    vlc_vector_clear (&p_sys->queue);
+    vlc_UrlClean (&p_sys->p_submit_url);
+
+    vlc_player_Lock (p_sys->player);
+    vlc_player_RemoveListener (p_sys->player , p_sys->player_listener);
+    vlc_player_Unlock (p_sys->player);
+
+    vlc_player_RemoveTimer (p_sys->player , p_sys->timer_listener);
+
+    free (p_sys);
+}
+
+static void *Run (void *data)
+{
+    intf_thread_t *p_intf = data;
+    int canc = vlc_savecancel ();
+    bool b_wait = 0;
+    char *request , *payload;
+
+    intf_sys_t *p_sys = p_intf->p_sys;
+
+    while (1)
+    {
+        vlc_restorecancel (canc);
+
+        if (b_wait)
+            vlc_tick_wait (vlc_tick_now () + VLC_TICK_FROM_SEC (60)); // wait for 1 min
+
+        vlc_mutex_lock (&p_sys->lock);
+        mutex_cleanup_push (&p_sys->lock) ;
+        while (p_sys->queue.size == 0)
+            vlc_cond_wait (&p_sys->wait , &p_sys->lock);
+        vlc_cleanup_pop ();
+
+        payload = PreparePayload (p_intf);
```

Comment:
I think vlc_memstream functions are cancellation point. This mean that PreparePayload() can be cancelled. In that case, it will result in a deadlock.
I don't recommend you to use thread cancellation for this module. Instead you should use something like that: https://git.videolan.org/?p=vlc.git;a=blobdiff;f=src/posix/timer.c;h=713a9133f5563c65e98ef0dc86bf3e4b9cbf20d6;hp=9acd51db8f68153eba363a71d6cdb6c00a926d57;hb=93b89a2181a6d76663db5dd81a04800282af126f;hpb=878b7247b18358443999535b7787205485e5a280

Reply:
I have made the changes but am not sure.

Reply:
Hello, you understood the issue but your fix is not correct. 

Explanation:
 - the bool live need to be protected by the mutex since it is read from the Run thread and written from the VLC one
 - You need to signal the Run thread that you changed the live variable
 - Every wait mechanisms used by the Run thread (vlc_tick_wait(),and (p_sys->queue.size == 0)) should check for this live bool too.
 - Use vlc_cond_timedwait() instead of vlc_tick_wait() in order to be immediately leave when the variable is signalled.

Here is the diff you should apply on top of your changes.

```
diff --git a/modules/misc/listenbrainz.c b/modules/misc/listenbrainz.c
index 4c04c8d8a2..cca9de35d4 100755
--- a/modules/misc/listenbrainz.c
+++ b/modules/misc/listenbrainz.c
@@ -462,7 +462,11 @@ static void Close (vlc_object_t *p_this)
     intf_thread_t *p_intf = ( intf_thread_t * ) p_this;
     intf_sys_t *p_sys = p_intf->p_sys;

+    vlc_mutex_lock(&p_sys->lock);
     p_sys->live = false;
+    vlc_cond_signal (&p_sys->wait);
+    vlc_mutex_unlock(&p_sys->lock);
+
     vlc_interrupt_kill(p_sys->interrupt);
     vlc_join (p_sys->thread , NULL);
     vlc_interrupt_destroy(p_sys->interrupt);
@@ -489,16 +493,25 @@ static void *Run (void *data)

     vlc_interrupt_set(p_sys->interrupt);

-    while (p_sys->live)
+    for (;;)
     {
+        vlc_mutex_lock (&p_sys->lock);
         if (b_wait)
-            vlc_tick_wait (vlc_tick_now () + VLC_TICK_FROM_SEC (60)); // wait for 1 min
+        {
+            vlc_tick_t deadline = vlc_tick_now() + VLC_TICK_FROM_SEC (60);
+            int ret = 0;
+            while (p_sys->live && ret == 0) // wait for 1 min
+                ret = vlc_cond_timedwait (&p_sys->wait , &p_sys->lock, deadline);
+        }

-        vlc_mutex_lock (&p_sys->lock);
-        mutex_cleanup_push (&p_sys->lock) ;
-        while (p_sys->queue.size == 0)
+        while (p_sys->live && p_sys->queue.size == 0)
             vlc_cond_wait (&p_sys->wait , &p_sys->lock);
-        vlc_cleanup_pop ();
+
+        if (!p_sys->live)
+        {
+            vlc_mutex_unlock (&p_sys->lock);
+            return NULL;
+        }

         payload = PreparePayload (p_intf);
         vlc_mutex_unlock (&p_sys->lock);
```

Reply:
Thanks for the diff. I have applied it and pushed the changes. Please take a look and let me know if its ready for submission.

Reply:
I think I spotted the main issues. It's ready for submission on the VLC ML. When I have time, I will also run your module with asan/tsan.

Reply:
I think I spotted the main issues. It's ready for submission on the VLC ML. When I have time, I will also run your module with asan/tsan.